### PR TITLE
Split Dhall.Parser in smaller modules

### PR DESCRIPTION
**Pull request by zudov**
This is a good thing on it's own, but beside that it's very likely to help with #443

### CODE REVIEW

**Comment by Gabriel439** [comment]
Thank you! :)

**Comment by Gabriel439** [comment]
@eskimor: I don't know if you meant it as a joke but just in case I want to remind you that the Dhall contributing guidelines require that people treat others with dignity and respect:

https://github.com/dhall-lang/dhall-lang/blob/master/.github/CONTRIBUTING.md#do-i-belong-here

File: src/Dhall/Parser.hs [hunk 1]
```diff
@@ -61,710 +62,9 @@ import qualified Text.Parser.Combinators
...
-
-    let indexed = do
-            _ <- Text.Parser.Char.char '@'
-            Text.Parser.Token.natural
-
-    n <- indexed <|> pure 0
-    return (V x n)
+import Dhall.Parser.Combinators
+import Dhall.Parser.Token
+import Dhall.Parser.Expression
```
Comment (Gabriel439) [inline src/Dhall/Parser.hs]: I think you need to add `Dhall.Parser.Expression` to version control because I don't see it in the set of changes

Comment (zudov) [reply src/Dhall/Parser.hs]: Oh yes, sorry, I've pushed it now (and also added the modules to cabal file).

Thanks for looking at it, it's still a work in progress, so not really ready for final review. Let's discuss it tomorrow.

**Comment by zudov** [comment]
@eskimor @Gabriel439 Sorry about it, we were siting next to each other during that "review" and somehow ended up at bad joke.

**Comment by FintanH** [comment]
Beautiful 👌

**Comment by eskimor** [comment]
> https://github.com/dhall-lang/dhall-lang/blob/master/.github/CONTRIBUTING.md#do-i-belong-here

@zudov knew about that, but he said it was ok as I just insulted his code and not him. In any case, I'll update my comment as the PR is of course not all crap ;-)

**Comment by eskimor** [comment]
Also I am going to be back to serious now, as I am on my way back home already. Have an awesome remaining Zurihac! See you guys at next awesome Haskell event!

File: src/Dhall/Parser/Combinators.hs [hunk 2]
```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE DeriveDataTypeable         #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RecordWildCards            #-}
+module Dhall.Parser.Combinators where
```
Comment (Gabriel439) [inline src/Dhall/Parser/Combinators.hs]: Maybe add a header to each module explaining the scope of what belongs in that module

**Comment by zudov** [comment]
To summarize: I moved a bunch of code around without making any actual changes (to avoid regressions). The structure now goes as follows:

- `Dhall.Parser` :: The public parser interface, pretty much all definitions are exported and documented.
- `Dhall.Parser.Expression` :: High-level parsers that output `Expr` (or `Chunks`)
- `Dhall.Parser.Token` :: A bunch of parsers and scanners that capture small pieces of Dhall syntax outputting data-types from which `Expr`s can be constructed. Deals with things like literals, reserved words, comments, labels, filepaths, etc 
- `Dhall.Parser.Combinators` :: Utulities that don't directly relate to Dhall's syntax, but only help defining the rest of the parser.

`Expression` and `Token` modules are still quite big, so essentially instead of having one kitchen sink for everything, we now have a compound kitchen sink (`Dhall.Parser`) that consists of big sink for high-level dishes (`Dhall.Parser.Expression`), smaller sink for smaller dishes and cups (`Dhall.Parser.Token`) and a container with some cleaning devices (`Dhall.Parser.Combinators`).

As a next step I want to split it into more meaningful modules groupped by expression types, e.g. a module for imports, a module for literals, a module for records, a module for primitives, etc. Current structure is a very good stepping stone for that since now I can simply move stuff from `Expression` module and pull whatever needed from `Token` along with it.

In the end, this is ready to be merged and I'll continue with further localization and refactoring.

File: src/Dhall/Parser/Combinators.hs (hunk 2 above)
Comment (zudov) [reply src/Dhall/Parser/Combinators.hs]: :+1: They all have it, except this one. Adding now.

Comment (zudov) [reply src/Dhall/Parser/Combinators.hs]: Added a missing one, and refined some others.

**Comment by Gabriel439** [comment]
@zudov: Sounds good!  I'll go ahead and merge this, then.  For future reference, you have the commit bit and therefore you have permission to merge your pull requests yourself if somebody else approves the change.  The only exception I can think of is for pull requests that are waiting on a matching change to the language standard and even then it's not a big deal if you accidentally merge them prematurely

The worst that happens is that I can always open another pull request proposing any modifications that I had in mind

**Pull request merged**